packages/hardhat-test-helpers/src/chai/matchers/bigNumber.ts

```diff
@@ -0,0 +1,264 @@
+    );
+  }
+}
+
+function override(
+  method: Methods,
+  name: string,
+  negativeName: string,
+  utils: Chai.ChaiUtils
+) {
+  return (_super: (...args: any[]) => any) =>
+    overwriteBigNumberFunction(method, name, negativeName, _super, utils);
+}
+
+function normalize(
+  source:
+    | number
+    | bigint
+    | typeof BNType
+    | typeof EthersBigNumberType
+    | typeof BigNumberJsType
+    | string
+): bigint {
+  if (isEthersBigNumber(source)) {
+    return BigInt(source.toString());
+  } else if (isBN(source)) {
+    return BigInt(source.toString());
+  } else if (isBigNumberJsBigNumber(source)) {
+    return BigInt(source.toString(10));
+  } else if (
+    typeof source === "string" ||
+    typeof source === "number" ||
+    typeof source === "bigint"
+  ) {
+    return BigInt(source);
+  } else {
+    throw new Error(`cannot convert ${typeof source} to BigNumber`);
+  }
+}
+
+function isBigNumber(source: any): boolean {
+  return (
+    typeof source === "bigint" ||
+    isEthersBigNumber(source) ||
+    isBN(source) ||
+    isBigNumberJsBigNumber(source)
+  );
+}
+
+function overwriteBigNumberFunction(
+  functionName: Methods,
+  readableName: string,
+  readableNegativeName: string,
+  _super: (...args: any[]) => any,
+  chaiUtils: Chai.ChaiUtils
+) {
+  return function (this: Chai.AssertionStatic, ...args: any[]) {
+    const [actualArg] = args;
+    const expectedFlag = chaiUtils.flag(this, "object");
+    if (chaiUtils.flag(this, "doLength") && isEthersBigNumber(actualArg)) {
+      // TODO: consider whether we really need this case (support for eg
+      // `expect('foo').to.have.length.of.at.least(BigInt(2));`, and if so then
+      // get it tested and generalize it to work with all the different
+      // BigNumbers.
+      _super.apply(this, [actualArg.toNumber()]);
+      return;
+    }
+    function compare(method: Methods, lhs: bigint, rhs: bigint): boolean {
+      if (method === "eq") {
+        return lhs === rhs;
+      } else if (method === "gt") {
+        return lhs > rhs;
+      } else if (method === "lt") {
+        return lhs < rhs;
+      } else if (method === "gte") {
+        return lhs >= rhs;
+      } else if (method === "lte") {
+        return lhs <= rhs;
+      } else {
+        throw new Error(`Unknown comparison operation ${method}`);
+      }
+    }
+    if (isBigNumber(expectedFlag) || isBigNumber(actualArg)) {
+      const expected = normalize(expectedFlag);
+      const actual = normalize(actualArg);
+      throwIfUnsafe(expectedFlag);
+      throwIfUnsafe(actualArg);
+      this.assert(
+        compare(functionName, expected, actual),
+        `expected ${expected} to ${readableName} ${actual}`,
+        `expected ${expected} to ${readableNegativeName} ${actual}`,
+        expected,
+        actual
```

Comment:
Using `expected` and `actual` here means that the error looks like this:

![image](https://user-images.githubusercontent.com/417134/158395237-92f5961f-20db-489b-9ee4-c577bd65cd75.png)

I wonder if the 3n and 4n will be confusing for users.

Reply:
Do something like https://github.com/NomicFoundation/hardhat/blob/bd1d6b971fb213d47ad62f556ef30ec58480155d/packages/hardhat-ethers/src/internal/index.ts#L20 to do a sort of "repr" (eg from Python), to have the error message be as depicted here but to have the "diff" (constructed from the `expected`/`actual` arguments on line 182,183) reflect the type of BigNumber object that was given in the input to the matcher.

Reply:
@alcuadrado and I discussed offline and decided that we should do this instead: https://github.com/NomicFoundation/hardhat/pull/2428/commits/c28827d1a2c404f15e813e5bb656932fcbe12c0a